include/deal.II/lac/solver_gcr.h

```diff
@@ -0,0 +1,202 @@
+            const AdditionalData &data = AdditionalData());
+
+  /**
+   *  Solve the linear system $Ax=b$ for x.
+   */
+  template <typename MatrixType, typename PreconditionerType>
+  void
+  solve(const MatrixType &        A,
+        VectorType &              x,
+        const VectorType &        b,
+        const PreconditionerType &preconditioner);
+
+private:
+  /**
+   * Maximum number of temporary vectors.
+   */
+  const unsigned int max_n_tmp_vectors;
+
+  /**
+   * Flag for the default residual that is used to measure convergence.
+   */
+  const bool use_default_residual;
+};
+
+
+
+template <class VectorType>
+inline SolverGCR<VectorType>::AdditionalData::AdditionalData(
+  const unsigned int max_n_tmp_vectors,
+  const bool         use_default_residual)
+  : max_n_tmp_vectors(max_n_tmp_vectors)
+  , use_default_residual(use_default_residual)
+{}
+
+
+
+template <class VectorType>
+inline SolverGCR<VectorType>::SolverGCR(SolverControl &       solver_control,
+                                        const AdditionalData &data)
+  : SolverBase<VectorType>(solver_control)
+  , max_n_tmp_vectors(data.max_n_tmp_vectors)
+  , use_default_residual(data.use_default_residual)
+{
+  solver_control.set_max_steps(max_n_tmp_vectors);
+}
+
+
+
+template <class VectorType>
+template <typename MatrixType, typename PreconditionerType>
+inline void
+SolverGCR<VectorType>::solve(const MatrixType &        A,
+                             VectorType &              x,
+                             const VectorType &        b,
+                             const PreconditionerType &preconditioner)
+{
+  using number = typename VectorType::value_type;
+
+  SolverControl::State conv = SolverControl::iterate;
+
+  typename VectorMemory<VectorType>::Pointer search_pointer(this->memory);
+  typename VectorMemory<VectorType>::Pointer Asearch_pointer(this->memory);
+  typename VectorMemory<VectorType>::Pointer p_pointer(this->memory);
+
+  VectorType &search  = *search_pointer;
+  VectorType &Asearch = *Asearch_pointer;
+  VectorType &p       = *p_pointer;
+
+  std::vector<typename VectorType::value_type> Hn_preloc(max_n_tmp_vectors);
+
+  internal::SolverGMRESImplementation::TmpVectors<VectorType> H_vec(
+    max_n_tmp_vectors, this->memory);
+  internal::SolverGMRESImplementation::TmpVectors<VectorType> Hd_vec(
+    max_n_tmp_vectors, this->memory);
+
+  search.reinit(x);
+  Asearch.reinit(x);
+  p.reinit(x);
+
+  typename VectorType::value_type res = 0.0;
+
+  A.vmult(p, x);
+  p.add(-1., b);
+
+  if (use_default_residual == true)
+    res = p.l2_norm();
+
+  preconditioner.vmult(search, p);
+
+  if (use_default_residual == false)
+    res = search.l2_norm();
+
+  unsigned int it = 0;
+
+  conv = this->iteration_status(it, res, x);
+  if (conv != SolverControl::iterate)
+    return;
+
+  while (conv == SolverControl::iterate)
+    {
+      it++;
+
+      const unsigned int it_ = ((it - 1) % max_n_tmp_vectors) + 1; // restart
```

Comment:
an attempt for restart if the number of iterations is higher than max_n_tmp_vectors